File: 08-07-25/559. Maximum Depth of N-ary Tree.kt
/**
 * Definition for a Node.
 * class Node(var `val`: Int) {
 *     var children: List<Node?> = listOf()
 * }
 */

class Solution {
    fun maxDepth(root: Node?): Int {
        if(root==null) return 0
         var maxD=0
         val stack=ArrayDeque<Pair<Node, Int>>()

        stack.add(Pair(root,1))

        while(stack.isNotEmpty()){
            val (node,depth)=stack.removeLast()
            maxD=maxOf(maxD,depth)
            for(child in node.children){
                if(child !=null)
                    stack.add(Pair(child,depth+1))
            }
        }
        return maxD
    }
}

File: 08-07-25/559. Maximum Depth of N-ary Tree.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    vector<Node*> children;

    Node() {}

    Node(int _val) {
        val = _val;
    }

    Node(int _val, vector<Node*> _children) {
        val = _val;
        children = _children;
    }
};
*/

class Solution {
public:
    int maxDepth(Node* root) {
        if(root==nullptr) return 0;

        int maxD=0;
        stack<pair<Node*, int>>st;

        st.push({root,1});
        while( !st.empty()){
            auto[node,depth]=st.top();
            st.pop();

            maxD=max(maxD,depth);

            for(auto child: node->children){
                if(child){
                    st.push({child,depth+1});
                }
            }
        }
        return maxD;
    }
};
